### PR TITLE
Added emms as doom module { :app emms }

### PR DESCRIPTION
**Pull request by tJ2ow**
My documentation isn't great

### CODE REVIEW

File: modules/app/emms/config.el [hunk 1]
```diff
@@ -0,0 +1,49 @@
+;;; private/emms/config.el -*- lexical-binding: t; -*-
+
+(use-package emms
+  :config
```
Comment (hlissner) [inline modules/app/emms/config.el:4]: The package is not lazy loaded. Please lazy load it.

File: modules/app/emms/config.el [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  (emms-all))
```
Comment (hlissner) [inline modules/app/emms/config.el:5]: The `use-package` block is prematurely closed here. i.e. there's one too many `)`.

File: modules/app/emms/config.el [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        :n "c" #'mpd/start-music-daemon
+        :n "k" #'mpd/kill-music-daemon
+        :n "u" #'mpc/update-database)
+  (map! :prefix ((concat doom-leader-alt-key " e ") . "emms")
+        :ie "p" #'emms-playlist-new
+        :ie "b" #'emms-smart-browse
+        :ie "r" #'emms-player-mpd-update-all-reset-cache
+        :ie "c" #'mpd/start-music-daemon
+        :ie "k" #'mpd/kill-music-daemon
+        :ie "u" #'mpc/update-database)
```
Comment (hlissner) [inline modules/app/emms/config.el:20]: Modules should not be binding leader keys, only localleader keys. Please remove these. At most, leader keys should be bound in the `:config default` module. That said, I won't accept `SPC e` as a leader key, so just remove them for now and I'll decide on one later.

File: modules/app/emms/config.el [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  (defun mpd/kill-music-daemon ()
+    (interactive)
+    (emms-stop)
+    (call-process "killall" nil nil nil "mpd")
+    (message "MPD Killed!"))
+
+  (defun mpc/update-database ()
+    (interactive)
+    (call-process "mpc" nil nil nil "update")
+    (message "MPD Database Updated!"))
```
Comment (hlissner) [inline modules/app/emms/config.el:49]: Move these to `apps/emms/autoload.el` and place `;;;###autoload` cookies above them.

File: modules/app/emms/init.el [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  (require 'emms-player-mpd)
+  (setq emms-directory (concat user-emacs-directory "/.local/emms"))
+  (setq emms-history-file (concat emms-directory "/history"))
+  (setq emms-cache-file (concat emms-directory "/cache"))
+  (setq emms-score-file (concat emms-directory "/score"))
+  (setq emms-seek-seconds 5)
+  (setq emms-player-list '(emms-player-mpd))
+  (setq emms-info-functions '(emms-info-mpd))
+  (setq emms-player-mpd-server-name "localhost")
+  (setq emms-player-mpd-server-port "8501"))
```
Comment (hlissner) [inline modules/app/emms/init.el:17]: Why are these in init.el? Move them to config.el. Please lazy load the package.

File: modules/app/emms/config.el [hunk 6]
```diff
@@ -0,0 +1,49 @@
+;;; private/emms/config.el -*- lexical-binding: t; -*-
```
Comment (hlissner) [inline modules/app/emms/config.el:1]: Should be apps/emms/config.el

File: modules/app/emms/packages.el [hunk 7]
```diff
@@ -0,0 +1,4 @@
+;; -*- no-byte-compile: t; -*-
+;;; private/emms/packages.el
+
+(package! emms)
```
Comment (hlissner) [inline modules/app/emms/packages.el:4]: Please pin this package: place your cursor in the `package!` statement and execute `M-x doom/bump-package-at-point`

File: modules/app/emms/packages.el [hunk 8]
```diff
@@ -0,0 +1,4 @@
+;; -*- no-byte-compile: t; -*-
+;;; private/emms/packages.el
```
Comment (hlissner) [inline modules/app/emms/packages.el:2]: Should be apps/emms/packages.el

File: modules/app/emms/config.el (hunk 4 above)
Comment (hlissner) [reply modules/app/emms/config.el:49]: And change their names to start with `+emms/mpd-*`, to conform to Doom's naming conventions.

File: modules/app/emms/README.org [hunk 9]
```diff
@@ -0,0 +1,85 @@
+#+TITLE:   app/emms
+#+DATE:    June 1, 2020
```
Comment (hlissner) [inline modules/app/emms/README.org:2]: Add `#+SINCE:   v3.0.0` below `#+DATE` line

File: modules/app/emms/README.org [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+* Description
+
+Emms is a music player written in EMACS-lisp.
+It uses mpd as a backend server, and mpc for update the music db.
+
+* Prerequisites
+- mkdir a dir in either $HOME/.mpd or $HOME/.config/mpd
+- make a file in mpd dir named mpd.conf
+#+BEGIN_QUOTE
```
Comment (hlissner) [inline modules/app/emms/README.org:22]: This should be wrapped in a `#+BEGIN_SRC conf` not a `#+BEGIN_QUOTE`.

Comment (hlissner) [reply modules/app/emms/README.org:22]: Please indent the `#+BEGIN_SRC` block with two spaces, to match the list item it's in.

File: modules/app/emms/README.org [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+- [[#description][Description]]
+- [[#prerequisites][Prerequisites]]
+  - [[#ubuntu][Ubuntu]]
+  - [[#arch-linux][Arch Linux]]
+- [[#troubleshooting][Troubleshooting]]
+  - [[#new-music-files-in-db-not-showing-up][new music files in db not showing up]]
+- [[#commands][Commands]]
+
+* Description
+
```
Comment (hlissner) [inline modules/app/emms/README.org:15]: Remove this blank line between heading and paragraph.

File: modules/app/emms/README.org [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+audio_output {
+        type "pulse"
+        name "pulse audio"
+}
+
+audio_output {
+    type                    "fifo"
+    name                    "my_fifo"
+    path                    "/tmp/mpd.fifo"
+    format                  "44100:16:2"
```
Comment (hlissner) [inline modules/app/emms/README.org:40]: Please make the indentation consistent.

Comment (hlissner) [inline modules/app/emms/README.org:40]: Reduce the amount of unnecessary whitespace between keys and values.

```diff
-    type                    "fifo"
-    name                    "my_fifo"
-    path                    "/tmp/mpd.fifo"
-    format                  "44100:16:2"
+    type    "fifo"
+    name    "my_fifo"
+    path    "/tmp/mpd.fifo"
+    format  "44100:16:2"
```

File: modules/app/emms/README.org [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+bind_to_address "127.0.0.1"
+port "8501"
+#+END_QUOTE
+- make the following files in mpd dir
+  mpdstate
+  mpd.log
```
Comment (hlissner) [inline modules/app/emms/README.org:48]: These should be a sublist.

```org
- make the following files in mpd dir
  - mpdstate
  - mpd.log
```

File: modules/app/emms/README.org [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+* Troubleshooting
+** new music files in db not showing up
+Try =SPC e k=  =SPC e c= =SPC e r=.
+If that doesn't work restart emacs.
+* Commands
+These are the most important ones
+
+=SPC e c= Updates DB and start mpd daemon.
+=SPC e k= Stop the player and kill the daemon.
+=SPC e b= Opens the emms browser move to the right window and the emms-mode keys will work.
```
Comment (hlissner) [inline modules/app/emms/README.org:70]: Remove these. See later comment re: modules shouldn't be binding leader keys.

File: modules/app/emms/README.org [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+** Arch Linux
+#+BEGIN_SRC bash
+sudo pacman -S mpd mpc
+#+END_SRC
+
+* Troubleshooting
+** new music files in db not showing up
+Try =SPC e k=  =SPC e c= =SPC e r=.
+If that doesn't work restart emacs.
```
Comment (hlissner) [inline modules/app/emms/README.org:64]: Please explain what these keybinds are doing (at the very least, what commands they're bound to).

File: modules/app/emms/README.org [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+=g a= Sort the browser by artists
+=g b= Sort the browser by genre
+=X= Stops the player
+=x= pauses the player
+=D= in the playlist window removes that item
+=C= clears the playlist
+=>= move five seconds forward in the current track
+=<= Opposite five back
+=RET= In browser adds the artist's songs to the playlist, in the playlist starts playlist.
```
Comment (hlissner) [inline modules/app/emms/README.org:85]: Convert these keys into a list by prefixing them with a `+`

File: modules/app/emms/config.el (hunk 4 above)
Comment (hlissner) [reply modules/app/emms/config.el:49]: The implementation is also very platform dependent (and will likely break on Windows, or systems where killall isn't installed.

The mpd process should be managed by Emacs with `make-process` and killed with `kill-process` instead.

File: modules/app/emms/config.el (hunk 1 above)
Comment (tJ2ow) [reply modules/app/emms/config.el:18]: I'm not sure if :defer t lazy loads or not.

**Review by hlissner — APPROVED** [review]

**Comment by hlissner** [comment]
Sorry for the _super_ late reply. With a few minor corrections this is ready for merging. Thanks for your help!
